src/sql/parts/notebook/models/notebookModel.ts

```diff
@@ -428,6 +428,11 @@ export class NotebookModel extends Disposable implements INotebookModel {
 
 	public doChangeKernel(kernelSpec: nb.IKernelSpec): Promise<void> {
 		this.setProviderIdForKernel(kernelSpec);
+		// Ensure that the kernel we try to switch to is a valid kernel; if not, use the default
+		if (this.notebookManager && this.notebookManager.sessionManager && this.notebookManager.sessionManager.specs &&
```

Comment:
OK. Let's add a helper function to getSpecs (which could be null), or even getKernelSpecs, then put the logic with && in one place. And then you can just do 
```suggestion
let kernelSpecs = this.getKernelSpecs();
if (kernelSpecs && kernelSpecs.findIndex(k => k.name === kernelSpec.name) < 0) {
```